---
title: "R Assignment"
output:
  html_document: default
  html_notebook: default
---

#R Assignment

Your R assignment will consist of two components:
1. Replicating your UNIX assignment in R
2. Additional analysis and visualization

Do your R assignment in your "homework" repository. Explain all your steps and include your code in an "R Notebook" file.  In addition, update the `README.md` file to indicate the extra work you have done.

## Part I
### _Data Inspection_

Load the two data files you used for your UNIX assignment in R and inspect their context. Use as many functions as you can to describe their structure and their dimensions (file size, number of columns, number of lines, ect...). You don't have to limit yourselves to the functions we learned in class.

As a reminder, the files are:

1. `fang_et_al_genotypes.txt`: a published SNP data set including maize, teosinte (i.e., wild maize), and Tripsacum (a close outgroup to the genus _Zea_) individuals


2. `snp_position.txt`: an additional data file that includes the SNP id (first column), chromosome location (third column), nucleotide location (fourth column) and other information for the SNPs genotyped in the `fang_et_al_genotypes.txt` file


```{r}
# Load library
library(tidyverse)
# Import genotype data
genotype <- read.table("fang_et_al_genotypes.txt", header=TRUE, stringsAsFactor = FALSE)
# Check file size
file.info("fang_et_al_genotypes.txt")
# Check number of columns, number of lines
dim(genotype)
nrow(genotype)
ncol(genotype)
# Import snp data
snp <- read.table("snp_position.txt", sep = "\t", header=TRUE, stringsAsFactor = FALSE)
# Check file size
file.info("snp_position.txt")
# Check number of columns, number of lines
dim(snp)
nrow(snp)
ncol(snp)
```

### _Data Processing_

Now we will manipulate the two files in R in order to format them for a downstream analysis. During this process, we will need to `join` (hint, hint) these data sets so that we have both genotypes and positions in a series of input files. All our files will be formatted such that the first column is "SNP_ID", the second column is "Chromosome", the third column is "Position", and subsequent columns are genotype data from either maize or teosinte individuals.

For maize (Group = ZMMIL, ZMMLR, and ZMMMR in the third column of the `fang_et_al_genotypes.txt` file) we want 20 files in total:

* 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

* 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -

For teosinte (Group = ZMPBA, ZMPIL, and ZMPJA in the third column of the `fang_et_al_genotypes.txt` file) we want 20 files in total:

* 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

* 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -

A total of 40 files will therefore be produced.

### A few notes and hints:
* In order to join these files, you may need to transpose your genotype data so the columns become rows.  You just have to know one letter to do this in R: `t()`.  However, check the results carefully, as there will be surprises ;)

* As in the UNIX assignment, it might help to write out the entire workflow that will be necessary to produce the files described above before doing the actual analysis.

* If you get stuck or confused, first, use the help() function; second, search the Internet; and, finally, post to the "scripting_help" channel on Slack and we will provide hints that may be helpful for the whole class.

```{r}
# Extracted the selected maize data and teosinte data
maize_genotype <- c('ZMMLR','ZMMMR','ZMMIL')
maize <- filter(genotype, Group %in% maize_genotype)
teosinte_genotype <- c('ZMPBA','ZMPJA','ZMPIL')
teosinte <- filter(genotype, Group %in% teosinte_genotype)
# Another way to do it: genotype_maize <- filter(genotype_tbl, Group =='ZMMLR'|Group =='ZMMMR'|Group == 'ZMMIL')

# Transpose and 
maize_t <- t(maize)
teosinte_t <- t(teosinte)

# Convert the transposed matrix into dataframe and remove top three rows of header
maize_t_trim <- data.frame(maize_t[-(1:3),],stringsAsFactors=FALSE)
teosinte_t_trim <- data.frame(teosinte_t[-(1:3),],stringsAsFactors=FALSE)

# Add row name as a new column named with SNP_ID
SNP_ID <- rownames(maize_t_trim)
rownames(maize_t_trim) <- NULL
maize_t_trim_name <- cbind(SNP_ID, maize_t_trim)

SNP_ID <- rownames(teosinte_t_trim)
rownames(teosinte_t_trim) <- NULL
teosinte_t_trim_name <- cbind(SNP_ID,teosinte_t_trim)

# For snp data, select the 1st, 3rd and 4th column, make SNP_ID to be factor, so that it can join
snp_select <- snp[c(1,3,4)]
snp_select$SNP_ID <- as.factor(snp_select$SNP_ID)

# Join snp data and maize genetope by their shared column SNPID
joined_maize <- left_join(snp_select, maize_t_trim_name, by =c("SNP_ID"))
joined_teosinte <- left_join(snp_select,teosinte_t_trim_name, by =c("SNP_ID"))

# Generate new files with missing data encoded by -
joined_maize_dash <- data.frame(lapply(joined_maize, gsub, pattern = "\\?", replacement = "\\-"))
joined_teosinte_dash <- data.frame(lapply(joined_teosinte, gsub, pattern = "\\?", replacement = "\\-"))

# Function for filtering chromosome and ordering SNP position increasingly, x is chromosome number, y is the dataframe to be used, filename is the identifier name in the file exported
sortChrIncrease <- function(x, y, filename) {
df <- y %>% filter(Chromosome == x) %>% arrange(as.numeric(Position))
write.table(df, paste0(filename,"sortChr",x,"_Increase.txt"), sep = "\t")
}

# Function for filtering chromosome and ordering SNP position decreasingly, x is chromosome number, y is the dataframe to be used, filename is the identifier name in the file exported
sortChrDecrease <- function(x, y, filename) {
df <- y %>% filter(Chromosome == x) %>% arrange(desc(as.numeric(Position)))
write.table(df, paste0(filename,"sortChr",x,"_Decrease.txt"), sep = "\t")
}

# Maize 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
for (i in 1:10) {
  sortChrIncrease(i, joined_maize, "Maize_")
}

# Maize 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
for (i in 1:10) {
  sortChrDecrease(i, joined_maize_dash, "Maizedash_")
}

# Teosinte 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
for (i in 1:10) {
  sortChrIncrease(i, joined_teosinte, "Teosinte_")
}

# Teosinte 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
for (i in 1:10) {
  sortChrDecrease(i, joined_teosinte_dash, "Teosintedash_")
}
```

### Part II
We will use ggplot to visualize our data in this part. As mentioned today in class, it may be easier to reshape the original data (make it tidy) using the melt command in the reshape2 package before attempting this part.
### SNPs per chromosome
Plot the total number of SNPs in our dataset on each chromosome. What groups contribute most of these SNPs?
```{r}
# Plot SNPs against chromosome
snp_per_chr <- snp %>% group_by (Chromosome) %>% summarise(snp_number = length(SNP_ID))

ggplot(data = snp_per_chr) +
  geom_bar(mapping = aes(x = Chromosome, y = snp_number), stat = "identity")

# Tidy data and plot SNPs against groups
library(reshape2)
genotype_long <- melt(genotype, id.var=c("Sample_ID", "Group"), measure.vars = 4:968,
                variable.name = "SNP_ID")
joined_genotype_long <- left_join(snp_select, genotype_long, by =c("SNP_ID"))
snp_per_Group <- joined_genotype_long %>% group_by (Group) %>% summarise(snp_number = length(SNP_ID))
ggplot(data = snp_per_Group) +
  geom_bar(mapping = aes(x = Group, y = snp_number), stat = "identity")
```

The chromosome 1, 2, 5, and 7 are enriched in SNPs.
ZMMLR and ZMPBA contribute most of these SNPs.


### Missing data and amount of heterozygosity
Create a new column to indicate whether a particular site is homozygous (has the same nucleotide on both chromosomes (i.e., A/A, C/C, G/G, T/T) or heterozygous (otherwise)). Recode the missing data as NA. Sort your dataframe using Group and Species_ID values. Make a graph that shows the proportion of homozygous and heterozygous sites as well as missing data in each species (you won't be able to see species names) and each group. For groups normalize the height of individual bars using one of the ggplot "position adjustments" options.
```{r}
# Find out the possibilities of heterozygosity
unique(joined_genotype_long$value)
# Replace NA to "missing", had problem pass NA to function
joined_genotype_long$value[which(is.na(joined_genotype_long$value))] <- "missing"
# Build a function to check heterozygosity
myfunction <- function(x){
  if (x == "A/A" |x == "C/C"|x == "G/G"|x == "T/T") {
  y <-  "Homozygous"
    } else if (x == "C/G"|x == "A/C"|x =="G/T"|x =="A/G"|x =="C/T"|x =="A/T") {
  y <-  "Heterozygous"
    }
    else
  y <- NA
  return(y)
}

# Create a new column for heterozygosity
heterozygosity <- sapply(joined_genotype_long$value[1:2684648],myfunction)

joined_genotype_long_heterozygosity <- cbind(joined_genotype_long[1:2684648,],heterozygosity)

# Plot the heterozygosity for each group
ggplot(data = joined_genotype_long_heterozygosity, aes(heterozygosity)) +
  geom_bar(aes(fill=Group), position = "fill")

# Note: loop is another option
# Create an empty column to save heterozygosity information  
#joined_genotype_long$heterozygosity <- NA
#for(i in 1:length(joined_genotype_long$heterozygosity)){ joined_genotype_long$heterozygosity[i] <- myfunction(joined_genotype_long$value[i])}

# Note2, cannot use is.na(NA) or x == "NA" in myfunction
```

### Your own visualization
Visualize one other feature of the dataset. The choice is up to you!

I am curious to see among these observations, the cases of homozygous and heterozygous sites as well as missing data in each chromosome.

```{r}
ggplot(data = joined_genotype_long_heterozygosity, aes(heterozygosity)) +
  geom_bar(aes(fill=Chromosome),position = "dodge")
```



